components/api_version/src/api_v2.rs

```diff
@@ -48,26 +51,31 @@ impl KvFormat for ApiV2 {
         }
 
         match key[0] {
-            RAW_KEY_PREFIX => KeyMode::Raw,
-            TXN_KEY_PREFIX => KeyMode::Txn,
+            RAW_KEY_PREFIX if key.len() >= KEY_PREFIX_SIZE => KeyMode::Raw,
+            TXN_KEY_PREFIX if key.len() >= KEY_PREFIX_SIZE => KeyMode::Txn,
             TIDB_META_KEY_PREFIX | TIDB_TABLE_KEY_PREFIX => KeyMode::TiDB,
             _ => KeyMode::Unknown,
         }
     }
 
     fn parse_range_mode(range: (Option<&[u8]>, Option<&[u8]>)) -> KeyMode {
-        match range {
-            (Some(start), Some(end))
-                if !start.is_empty()
-                    && !end.is_empty()
-                    && (start[0] == end[0] ||
-                        // Special case to represent "".."" within a key mode
-                        (end == [start[0] + 1])) =>
+        if let (Some(start), Some(end)) = range {
+            if start.is_empty() || end.is_empty() {
+                return KeyMode::Unknown;
+            }
+
+            let mode = Self::parse_key_mode(start);
+
+            if mode == KeyMode::Unknown
+                || start[0] == end[0]
+                || (mode == KeyMode::TiDB && end == [start[0] + 1])
```

Comment:
If we combine them into one function, requiring that `infinite_end_of` could return both `[u8; 1] or [u8; 4]`, this might result in allocating a `Vec` instead of a fix-sized array.